[JSC] Add more tests for RegExp V flag and fix issues found

msaboff opened the pull request:
#### 30c8a2c9c4a8874d97236588a3241ca247e8fe36
<pre>
[JSC] Add more tests for RegExp V flag and fix issues found
<a href="https://bugs.webkit.org/show_bug.cgi?id=253915">https://bugs.webkit.org/show_bug.cgi?id=253915</a>
rdar://106724836

Reviewed by Yusuke Suzuki.

Added over 40 tests for RegExp V flag, organizing the syntax error tests ahead of the operational tests.

This change doesn&apos;t impact the results of the test262 RegExp V flag related tests.  Rather it fills in
the missing syntax checks from test262.

Almost all of the issues found with these new tests impact the ClassSetParserDelegate processing.
These changes include:
 * Saving and restoring the current parsing state when nesting class sets.  That change allowed
   eliminating the nesting class level and using the size of the nested parse state vector.
 * Implicitly change the opcode type from Default to Union when we have a construct that is only
   allowed for the ClassUnion production.  For example, a simple unnested rangecl like a-z cannot appear
   as part of a ClassIntersection or ClassSubtraction production.  This is done with the new method
   switchFromDefaultOpToUnionOpIfNeeded().
 * Added explicit state handling for set operands.  This facilitates the proper processing of
   ClassSetOperands per the spec.
 * Added syntax checking for the ClassSetSyntaxCharacter and ClassSetReservedPunctuator rules
   according to the spec.
 * Renamed the CharacterClassConstructionState enum in ClassSetParserDelegate to ClassSetConstructionState.
 * Renamed doneAfterCharacterClassEnd() to nestedClassEnd() to match nestedClassBegin().

Since the return value of parseClassSetEscape() wasn&apos;t used, changed the type to void.

Updated CharacterClassConstructor::putChar(UChar32) with a new helper to handle adding characters
as part of the right hand side of a intersection or subtraction set operation.

Added a new error, InvalidClassSetCharacter, and changed the spelling of all the errors added with the
RegExp V flag initial checkin to have an initial lower case letter like all the other RegExp error strings.

* JSTests/stress/regexp-vflag-property-of-strings.js:
(testRegExpSyntaxError):
* Source/JavaScriptCore/yarr/YarrErrorCode.cpp:
(JSC::Yarr::errorMessage):
(JSC::Yarr::errorToThrow):
* Source/JavaScriptCore/yarr/YarrErrorCode.h:
* Source/JavaScriptCore/yarr/YarrParser.h:
(JSC::Yarr::Parser::ClassSetParserDelegate::NestingState::NestingState):
(JSC::Yarr::Parser::ClassSetParserDelegate::ClassSetParserDelegate):
(JSC::Yarr::Parser::ClassSetParserDelegate::nestedClassBegin):
(JSC::Yarr::Parser::ClassSetParserDelegate::nestedClassEnd):
(JSC::Yarr::Parser::ClassSetParserDelegate::switchFromDefaultOpToUnionOpIfNeeded):
(JSC::Yarr::Parser::ClassSetParserDelegate::setSubtractOp):
(JSC::Yarr::Parser::ClassSetParserDelegate::setIntersectionOp):
(JSC::Yarr::Parser::ClassSetParserDelegate::flushCachedCharacterIfNeeded):
(JSC::Yarr::Parser::ClassSetParserDelegate::afterOperand):
(JSC::Yarr::Parser::ClassSetParserDelegate::canTakeSetOperand):
(JSC::Yarr::Parser::ClassSetParserDelegate::setProcessingEscape):
(JSC::Yarr::Parser::ClassSetParserDelegate::atomPatternCharacter):
(JSC::Yarr::Parser::ClassSetParserDelegate::atomBuiltInCharacterClass):
(JSC::Yarr::Parser::ClassSetParserDelegate::end):
(JSC::Yarr::Parser::isIdentityEscapeAnError):
(JSC::Yarr::Parser::parseClassSetEscape):
(JSC::Yarr::Parser::parseClassSet):
(JSC::Yarr::Parser::ClassSetParserDelegate::doneAfterCharacterClassEnd): Deleted.
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::CharacterClassConstructor::putChar):
(JSC::Yarr::CharacterClassConstructor::putCharNonUnion):
(JSC::Yarr::CharacterClassConstructor::isUnionSetOp):

Canonical link: <a href="https://commits.webkit.org/261714@main">https://commits.webkit.org/261714@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a2d5f2adb8f532478b2f463307c057b89635a55

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-16-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/WinCairo-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-16-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-9-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8171 "Built successfully and passed tests") | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-9-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/yarr/YarrParser.h:
```diff
@@ -330,20 +342,26 @@ class Parser {
+            NestingState& lastState = nestedParseState.last();
+            m_setOp = lastState.m_setOp;
+            m_inverted = lastState.m_inverted;
+            nestedParseState.removeLast();
```
Let's use `takeLast()`.

msaboff replied:
Will change.

Constellation commented on Source/JavaScriptCore/yarr/YarrParser.h:
```diff
@@ -1056,14 +1186,23 @@ class Parser {
+
+            // Check if the character is part of ClassSetSyntaxCharacter, except the characters
+            // explicitly handled below or in classSetConstructor.atomPatternCharacter().
+            if (strchr("(){}/|)", ch) || !ch) {
```
Is this correct even for non ASCII ch?

msaboff replied:
I'll add an isASCII() check.

Constellation commented on Source/JavaScriptCore/yarr/YarrParser.h:
```diff
@@ -595,12 +722,15 @@ class Parser {
     {
-        if (isEitherUnicodeCompilation() && (!strchr("^$\\.*+?()[]{}|/", ch) || !ch)) {
+        if (isEitherUnicodeCompilation()
+            && (!strchr(isUnicodeCompilation() ? "^$\\.*+?()[]{}|/" : "^$\\.*+?()[]{}|/&-!#%,:;<=>@`~" , ch) || !ch)) {
```
Also, is it correct for non ASCII ch?

msaboff replied:
Part of that is existing code.  But to be safe, I'll add an isASCII() check.

Constellation reviewed (approved):
r=me with comments.

msaboff committed:
[JSC] Add more tests for RegExp V flag and fix issues found
https://bugs.webkit.org/show_bug.cgi?id=253915
rdar://106724836

Reviewed by Yusuke Suzuki.

Added over 40 tests for RegExp V flag, organizing the syntax error tests ahead of the operational tests.

This change doesn't impact the results of the test262 RegExp V flag related tests.  Rather it fills in
the missing syntax checks from test262.

Almost all of the issues found with these new tests impact the ClassSetParserDelegate processing.
These changes include:
 * Saving and restoring the current parsing state when nesting class sets.  That change allowed
   eliminating the nesting class level and using the size of the nested parse state vector.
 * Implicitly change the opcode type from Default to Union when we have a construct that is only
   allowed for the ClassUnion production.  For example, a simple unnested rangecl like a-z cannot appear
   as part of a ClassIntersection or ClassSubtraction production.  This is done with the new method
   switchFromDefaultOpToUnionOpIfNeeded().
 * Added explicit state handling for set operands.  This facilitates the proper processing of
   ClassSetOperands per the spec.
 * Added syntax checking for the ClassSetSyntaxCharacter and ClassSetReservedPunctuator rules
   according to the spec.
 * Renamed the CharacterClassConstructionState enum in ClassSetParserDelegate to ClassSetConstructionState.
 * Renamed doneAfterCharacterClassEnd() to nestedClassEnd() to match nestedClassBegin().

Since the return value of parseClassSetEscape() wasn't used, changed the type to void.

Updated CharacterClassConstructor::putChar(UChar32) with a new helper to handle adding characters
as part of the right hand side of a intersection or subtraction set operation.

Added a new error, InvalidClassSetCharacter, and changed the spelling of all the errors added with the
RegExp V flag initial checkin to have an initial lower case letter like all the other RegExp error strings.

* JSTests/stress/regexp-vflag-property-of-strings.js:
(testRegExpSyntaxError):
* Source/JavaScriptCore/yarr/YarrErrorCode.cpp:
(JSC::Yarr::errorMessage):
(JSC::Yarr::errorToThrow):
* Source/JavaScriptCore/yarr/YarrErrorCode.h:
* Source/JavaScriptCore/yarr/YarrParser.h:
(JSC::Yarr::Parser::ClassSetParserDelegate::NestingState::NestingState):
(JSC::Yarr::Parser::ClassSetParserDelegate::ClassSetParserDelegate):
(JSC::Yarr::Parser::ClassSetParserDelegate::nestedClassBegin):
(JSC::Yarr::Parser::ClassSetParserDelegate::nestedClassEnd):
(JSC::Yarr::Parser::ClassSetParserDelegate::switchFromDefaultOpToUnionOpIfNeeded):
(JSC::Yarr::Parser::ClassSetParserDelegate::setSubtractOp):
(JSC::Yarr::Parser::ClassSetParserDelegate::setIntersectionOp):
(JSC::Yarr::Parser::ClassSetParserDelegate::flushCachedCharacterIfNeeded):
(JSC::Yarr::Parser::ClassSetParserDelegate::afterOperand):
(JSC::Yarr::Parser::ClassSetParserDelegate::canTakeSetOperand):
(JSC::Yarr::Parser::ClassSetParserDelegate::setProcessingEscape):
(JSC::Yarr::Parser::ClassSetParserDelegate::atomPatternCharacter):
(JSC::Yarr::Parser::ClassSetParserDelegate::atomBuiltInCharacterClass):
(JSC::Yarr::Parser::ClassSetParserDelegate::end):
(JSC::Yarr::Parser::isIdentityEscapeAnError):
(JSC::Yarr::Parser::parseClassSetEscape):
(JSC::Yarr::Parser::parseClassSet):
(JSC::Yarr::Parser::ClassSetParserDelegate::doneAfterCharacterClassEnd): Deleted.
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::CharacterClassConstructor::putChar):
(JSC::Yarr::CharacterClassConstructor::putCharNonUnion):
(JSC::Yarr::CharacterClassConstructor::isUnionSetOp):

Canonical link: https://commits.webkit.org/261714@main
